c10/cuda/CUDACachingAllocator.cpp

```diff
@@ -728,38 +679,38 @@ class THCCachingAllocator {
     }
   }
 
-  void synchronize_and_free_events(optional<int> device) {
-    // Synchronize on outstanding events and then free associated blocks.
-    // Limited to blocks on the given device if specified.
+  cudaEvent_t create_event_internal() {
+    cudaEvent_t event;
+    if (cuda_event_cache.empty()) {
+      C10_CUDA_CHECK(cudaEventCreateWithFlags(&event, cudaEventDisableTiming));
+    } else {
+      event = cuda_event_cache.back();
+      cuda_event_cache.pop_back();
+    }
+    return event;
+  }
 
-    auto remaining_events = decltype(cuda_events)();
+  void free_event_internal(cudaEvent_t event) {
+    cuda_event_cache.push_back(event);
```

Comment:
As far as I can tell, this means cudaEventDestroy is ever called on the events, not even at program exit.

Reply:
Right, I took a cue from the fact that the cached allocations are never freed to mean that these objects were not intended to support fully leak-proof destruction.  This is also true for the streams created in `CUDAStream.cpp`.

Reply:
> the cached allocations are never freed 
wdym?  do you mean the tensors that remain live at the end of a training script are never explicitly freed?

Reply:
The tensors are freed, but the device memory underlying their storage is just placed back into the cache.  They are not freed back to the device proper until process exit (or `emptyCache` is invoked).